Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,347 @@
+/-
+Copyright (c) 2023 Adrian Wüthrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrian Wüthrich
+-/
+import Mathlib.Combinatorics.SimpleGraph.Basic
+import Mathlib.Combinatorics.SimpleGraph.DegreeSum
+import Mathlib.Combinatorics.SimpleGraph.Connectivity
+import Mathlib.Combinatorics.SimpleGraph.AdjMatrix
+import Mathlib.Combinatorics.SimpleGraph.IncMatrix
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Matrix.Rank
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Fintype.Card
+import Mathlib.Data.Real.Basic
+import Mathlib.LinearAlgebra.Basic
+import Mathlib.LinearAlgebra.Matrix.BilinearForm
+import Mathlib.LinearAlgebra.Finrank
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.IsDiag
+
+/-!
+# Laplacian Matrix
+
+This module defines the Laplacian matrix of a graph, and proves some of its elementary properties.
+
+## Main definitions & Results
+
+* `SimpleGraph.degMatrix`: The degree matrix of a simple graph
+* `SimpleGraph.lapMatrix `: The Laplacian matrix of a simple graph, defined as the difference
+  between the degree matrix and the adjacency matrix.
+* `isPosSemidef_lapMatrix`: The Laplacian matrix is positive semidefinite.
+* `rank_ker_lapMatrix_eq_card_ConnectedComponent`: The number of connected components in `G` is
+  the dimension of the nullspace of its Laplacian matrix.
+
+-/
+
+
+open BigOperators Finset Matrix SimpleGraph
+
+variable {V : Type*} [Fintype V] [DecidableEq V] (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- The diagonal matrix consisting of the degrees of the vertices in the graph. -/
+def SimpleGraph.degMatrix (R : Type*) [Ring R] : Matrix V V R := Matrix.diagonal (G.degree ·)
```

Comment:
I'm not sure if `[Ring R]` is the best assumption here, generally we go as weak as possible, so I'd it least try and include the naturals, but this should make sense for any `AddMonoidWithOne` I hope